client/components/tooltip/tooltip-base.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, useRef, useState, memo } from 'react';
+import { createPortal } from 'react-dom';
+import classNames from 'classnames';
+import { debounce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import './styles.scss';
+
+const rootElement =
+	document.getElementById( 'wpbody-content' ) || document.body;
+
+const isEventTriggeredWithin = ( event, element ) =>
+	element && ( element === event.target || element.contains( event.target ) );
+
+const useHideDelay = (
+	isVisibleProp,
+	{ hideDelay = 1000, triggerRef, tooltipRef, onClose }
+) => {
+	const [ isVisible, setIsVisible ] = useState( isVisibleProp );
+	// not using state for this, we don't need to cause a re-render
+	const hasMountedRef = useRef( false );
+
+	// hide delay
+	useEffect( () => {
+		let timer = null;
+
+		if ( ! hasMountedRef.current ) {
+			hasMountedRef.current = true;
+			return;
+		}
+
+		// element is marked as visible, no need to hide it
+		if ( isVisibleProp ) {
```

Comment:
I think this will be called if `isVisibleProp` has been true so far but you pass a different `onClose` handler or `hideDelay` value.

```suggestion
		if ( isVisibleProp && ! isVisible ) {
```

Reply:
Ideally the `onClose` (not `onHide`) handler _should_ maintain a stable reference.

That's a valid concern, tho - I put it in a `useRef` reference in 657dd84c , so that `onHide` can be removed from the dependencies 👍